[GOFF] Refactor writing GOFF records

redstar opened the pull request:
A GOFF file is basically a sequence of records. Change the YAML mapping for GOFF files to reflect this. As result, creating files for testing, e.g. without a module header, becomes possible.

redstar committed:
[GOFF] Refactor writing GOFF records

A GOFF file is basically a sequence of records. Change the YAML mapping for GOFF files to reflect this. As result, creating files for testing, e.g. without a module header, becomes possible.

redstar committed:
Replace combination of sed and bash expression with awk in LIT test

redstar committed:
Change awk statement  in LIT test again

jh7370 commented on llvm/lib/ObjectYAML/GOFFEmitter.cpp:
```diff
@@ -62,119 +63,90 @@ ZerosImpl zeros(const size_t NumBytes) { return ZerosImpl{NumBytes}; }
+  // The flags are determined by the flags of the prvious record, and by the
+  // remaining size of data.
+  uint8_t Flags = 0;
+  size_t Ptr = 0;
```
Could we avoid using `Ptr` for non pointer variables, please. I suggest `Pos` instead.

redstar replied:
Sure, changed.

jh7370 commented on llvm/lib/ObjectYAML/GOFFEmitter.cpp:
```diff
@@ -62,119 +63,90 @@ ZerosImpl zeros(const size_t NumBytes) { return ZerosImpl{NumBytes}; }
+  uint8_t Flags = 0;
+  size_t Ptr = 0;
+  size_t Size = Str.size();
+  while (Size >= GOFF::RecordContentLength) {
+    if (Flags) {
+      Flags |= Rec_Continuation;
+      if (Size == GOFF::RecordContentLength)
+        Flags &= ~Rec_Continued;
+    } else
+      Flags |= (Size == GOFF::RecordContentLength) ? 0 : Rec_Continued;
+    writeRecordPrefix(Flags);
+    OS.write(&Str.data()[Ptr], GOFF::RecordContentLength);
+    Size -= GOFF::RecordContentLength;
+    Ptr += GOFF::RecordContentLength;
   }
+  if (Size) {
+    Flags &= ~Rec_Continued;
+    writeRecordPrefix(Flags);
+    OS.write(&Str.data()[Ptr], Size);
+    OS.write_zeros(GOFF::RecordContentLength - Size);
+  }
```
This all feels a bit tricky to follow, and I wonder if an alternative implementation would be better? As I understand it, there are two main complexities that need resolving: calculating the flags and determining what to write. How about the following:

```
size_t Size = Str.size();
bool Continuation = false;
while (Size > 0) {
  uint8_t Flags = 0;
  if (Continuation)
    Flags |= Rec_Continuation;
  if (Size > GOFF::RecordContentLength) {
    Flags |= Rec_Continued;
    Continuation = true;
  }
  writeRecordPrefix(Flags);
  OS.write(&Str.data()[Ptr], Size);
  if (Size < GOFF::RecordContentLength) {
    OS.write_zeros(GOFF::RecordContentLength - Size);
    Size = 0;
  } else {
    Size -= GOFF::RecordContentLength;
  }
}
```

redstar replied:
Yes, your understanding is correct. I took your version with some small changes, thanks.

jh7370 commented on llvm/lib/ObjectYAML/GOFFEmitter.cpp:
```diff
@@ -185,81 +157,68 @@ class GOFFState {
+  SmallString<16> EntryName;
+  if (std::error_code EC =
+          ConverterEBCDIC::convertToEBCDIC(EndMod.EntryName, EntryName))
+    reportError("Conversion error on " + EndMod.EntryName);
```
Please refer to the coding standards for how to format error messages. Also applies below in various places.

redstar replied:
Changed all places.

jh7370 commented on llvm/lib/ObjectYAML/GOFFEmitter.cpp:
```diff
@@ -62,119 +63,90 @@ ZerosImpl zeros(const size_t NumBytes) { return ZerosImpl{NumBytes}; }
-    }
+void GOFFOStream::writeRecordPrefix(uint8_t Flags) {
+  uint8_t TypeAndFlags = Flags | (CurrentType << 4);
+  OS << binaryBe(static_cast<unsigned char>(GOFF::PTVPrefix))
+     << binaryBe(static_cast<unsigned char>(TypeAndFlags))
+     << binaryBe(static_cast<unsigned char>(0));
```
Can you use `uint8_t(...)` instead of the `static_cast`? This is for consistency with code you've written below.

jh7370 commented on llvm/lib/ObjectYAML/GOFFEmitter.cpp:
```diff
@@ -185,81 +157,68 @@ class GOFFState {
-  if (HasError)
-    return false;
-  writeEnd();
+  for (auto &RecPtr : Doc.Records) {
```
I don't think you should use `auto` here, based on my understanding of the coding standards: the type of the variable isn't immediately obvious from the RHS of the statement. Same immediately below.

Also, should it be `const`?

redstar replied:
Changed by removing auto, and using const.

jh7370 commented on llvm/lib/ObjectYAML/GOFFEmitter.cpp:
```diff
@@ -185,81 +157,68 @@ class GOFFState {
+    case GOFFYAML::RecordBase::RBK_Symbol:
+    case GOFFYAML::RecordBase::RBK_Text:
+    case GOFFYAML::RecordBase::RBK_DeferredLength:
+      llvm_unreachable(("Not yet implemented"));
```
Too many parentheses?

redstar replied:
Yes. :-( Changed.

jh7370 reviewed (commented):
I've reviewed the code. Will check the tests another time.

redstar committed:
Changes based on reviewer comments.

redstar commented on llvm/include/llvm/ObjectYAML/GOFFYAML.h:
```diff
@@ -25,25 +25,87 @@ namespace llvm {
+// forward references to records, etc.). However, to be able to specify invalid
+// GOFF files, we treat all records the same way.
+struct RecordBase {
+  enum RecordBaseKind {
```
Changed. I also renamed the type and the member., because I think that `RecordBase::Kind::ModuleHeader` is more readable than `RecordBase::RecordBaseKind::ModuleHeader`.

redstar commented on llvm/lib/ObjectYAML/GOFFEmitter.cpp:
```diff
@@ -62,119 +63,90 @@ ZerosImpl zeros(const size_t NumBytes) { return ZerosImpl{NumBytes}; }
-
-  ~GOFFOstream() { finalize(); }
+  explicit GOFFOStream(raw_ostream &OS)
+      : OS(OS), CurrentType(GOFF::RecordType(-1)) {}
```
Removed, since there is no need for a special magic number.

redstar commented on llvm/lib/ObjectYAML/GOFFEmitter.cpp:
```diff
@@ -62,119 +63,90 @@ ZerosImpl zeros(const size_t NumBytes) { return ZerosImpl{NumBytes}; }
+
+void GOFFOStream::write(StringRef Str) {
+  // The flags are determined by the flags of the prvious record, and by the
+  // remaining size of data.
```
Changed, thanks!

redstar committed:
Remove additional braces around argument list

Everybody0523 commented on llvm/test/tools/yaml2obj/GOFF/end-long-name.yaml:
```diff
@@ -0,0 +1,74 @@
+# CHECK1-EMPTY:
+
+## 1 record fully used.
+## Flags = 0 => no continuation, not continued.
```
```suggestion
## Flags = 0 => not a continuation, not continued.
```
Maybe it's just me but when I read "no continuation" I'm not sure whether it means the record in question has no continuation or that the record itself is not a continuation.

redstar replied:
Changed, thanks!

Everybody0523 commented on llvm/include/llvm/ObjectYAML/GOFFYAML.h:
```diff
@@ -25,25 +25,87 @@ namespace llvm {
+// forward references to records, etc.). However, to be able to specify invalid
+// GOFF files, we treat all records the same way.
+struct RecordBase {
+  enum class Kind {
```
Could you just use the RecordType enum from `llvm/BinaryFormat/GOFF.h` here?

redstar replied:
Those are currently the same, but I do not think that it is good to couple the record type with the class hierarchy discriminator. E.g. in C++ I could model a relocation entry as a record, and treat it as a sub-record of the relocation directory. I currently do not do this, but coupling the discriminator to the record type would be the design decision that I will never do that, which seems a bit limiting to me.

redstar committed:
Small changes

- Change comment in test case based on reviewer comment
- Add a test case testing error message for conversion error

jh7370 reviewed (commented):
Still not got to the tests in depth yet, but these might be better reviewed by somebody with GOFF expertise anyway.

redstar committed:
Updates based on reviewer comments

- Reword error message
- Put quotes around value in error message

redstar commented on llvm/test/tools/yaml2obj/GOFF/end-invalid-name.yaml:
```diff
@@ -0,0 +1,7 @@
+
+--- !GOFF
+- End:
+    EntryName: "Euro symbol €"
```
I agree with your worries. However, EBCDIC-1047 is designed to have all characters from ISO Latin-1. As result, to trigger this error message I need to use a character which is not in ISO Latin-1.
There are already LIT tests in clang which use characters requiring UTF-8, so it should be ok.

redstar commented on llvm/lib/ObjectYAML/GOFFEmitter.cpp:
```diff
@@ -181,11 +184,13 @@ void GOFFState::writeHeader(GOFFYAML::ModuleHeader &ModHdr) {
           ConverterEBCDIC::convertToEBCDIC(EndMod.EntryName, EntryName))
-    reportError("Conversion error on " + EndMod.EntryName);
+    reportError("conversion to EBCDIC 1047 failed on '" + EndMod.EntryName +
+                "'");
```
Changed, thanks.

redstar commented on llvm/lib/ObjectYAML/GOFFYAML.cpp:
```diff
@@ -71,26 +71,26 @@ void CustomMappingTraits<GOFFYAML::RecordPtr>::inputOne(
   } else if (Key == "RelocationDirectory" || Key == "Symbol" || Key == "Text" ||
              Key == "Length")
-    IO.setError(Twine("Not yet implemented ").concat(Key));
+    IO.setError(Twine("not yet implemented ").concat(Key));
```
Changed both places.

EsmeYi commented on llvm/lib/ObjectYAML/GOFFEmitter.cpp:
```diff
@@ -185,81 +158,71 @@ class GOFFState {
+  LR << binaryBe(uint8_t(EndMod.Flags)) // The flags.
+     << binaryBe(uint8_t(EndMod.AMODE)) // The addressing mode.
+     << zeros(3)                        // Reserved.
+     << binaryBe(EndMod.RecordCount)    // The record count.
```
If the record count is not specified, I think it would be better to fill in an actual calculated value.

kpneal replied:
I've run into trouble with the record count field. If it is zero, and the binder abends when trying to do a link, then I can use the Unix "dd" command to slice up the object and try and narrow down what is causing the abend. If the record count is correct to start with then slicing up the object will make it incorrect and remove a tool from my toolbox which then made my life more difficult.

EsmeYi replied:
To do this, you can specify the record count field as zero in yaml. And what I mean is when the field is omitted, i.e. has no value (neither zero nor any other value).

kpneal replied:
Correct, fields that are missing are filled with zero bytes in GOFF. A count of zero means the record count is unspecified.

redstar replied:
All non-specified fields are set to zero. (That's in GOFFYAML.cpp). I removed setting the block counter automatically because it would be a value not specified in the input file. If this is useful then I can add a special label to use in the yaml file.

jh7370 replied:
With the ELF version of yaml2obj, the intent of the calculated fields is so that minimal YAML can be used to create a valid ELF for testing purposes. In many places, a 0 value as the default makes sense though. This helps keep the signal-to-noise ratio in the tests down to a minimum. The aim should be the same here, so use 0 where a real value isn't needed for tools to be able to read the object, and use a calculated field where it is needed.

EsmeYi commented on llvm/lib/ObjectYAML/GOFFYAML.cpp:
```diff
@@ -12,34 +12,92 @@
+  ECase(AMODE_None);
+  ECase(AMODE_24);
+  ECase(AMODE_31);
+  ECase(AMODE_ANY);
```
Can it be changed to `AMODE_Any`? It is for consistency with `AMODE_None`.

kpneal replied:
No. AMODE_None means unspecified, and that means it defaults to AMODE 24. AMODE_ANY means either AMODE 24 or AMODE 31. These are different bit patterns in the object and thus cannot be collapsed without throwing away data from the object.

EsmeYi replied:
OK, I see. Thanks.

EsmeYi commented on llvm/lib/ObjectYAML/GOFFEmitter.cpp:
```diff
@@ -185,81 +158,71 @@ class GOFFState {
+  LogicalRecord LR(GW);
+  LR << zeros(45)                          // Reserved.
+     << binaryBe(ModHdr.ArchitectureLevel) // The architecture level.
+     << binaryBe(ModHdr.PropertiesLength)  // Length of module properties.
```
When this field is omitted in yaml, would it be better to fill it with a calculated value?

EsmeYi commented on llvm/lib/ObjectYAML/GOFFEmitter.cpp:
```diff
@@ -185,81 +158,71 @@ class GOFFState {
+     << binaryBe(EndMod.ESDID)          // ESDID of the entry point.
+     << zeros(4)                        // Reserved.
+     << binaryBe(EndMod.Offset)         // Offset of entry point.
+     << binaryBe(EndMod.NameLength)     // Length of external name.
```
Ditto. It can be filled with a calculated value when it's omitted.

kpneal replied:
Same comment as above. And I don't think the name field can ever be zero. A private symbol gets the name " " (a single space) which the binder then turns into something specific to that part of the GOFF file. I've since forgotten the details, but they don't matter here.

redstar commented on llvm/lib/ObjectYAML/GOFFYAML.cpp:
```diff
@@ -12,34 +12,92 @@
 void MappingTraits<GOFFYAML::Object>::mapping(IO &IO, GOFFYAML::Object &Obj) {
   IO.mapTag("!GOFF", true);
-  IO.mapRequired("FileHeader", Obj.Header);
+  EmptyContext Context;
```
Yes, but the purpose of yaml2obj is create test data, which also needs to include wrong data. E.g. if I require the module header to be present, I cannot write a lit test for the file reader, which checks that the module header comes first.

redstar replied:
Btw, being not flexible enough to produce test data containing wrong constellations was a big review issue when Yusra came up with a follow-up PR adding more functionality. I agree with that argument, since I this tool is mainly used to generate input data for the other MC tools.

EsmeYi replied:
Ok, fair enough, thanks for explanation.

kpneal commented on llvm/test/tools/yaml2obj/GOFF/end-long-name.yaml:
```diff
@@ -0,0 +1,74 @@
+## 1 record fully used.
+## Flags = 0 => not a continuation, not continued.
+# CHECK2:      03 40 00 00 00 00 00 00 00 00 00 00 00 00 00 00
+# CHECK2-NEXT: 00 00 00 00 00 00 00 00 00 00 c6 96 96 c2 81 99
```
The Name Length field is zero. From the documentation: "This length cannot be zero if a name is present."

Either we're writing invalid GOFF or the documentation needs an update.

kpneal replied:
Or perhaps a comment in the test that says we're purposely ignoring the name length field for this test.

kpneal commented on llvm/test/tools/yaml2obj/GOFF/end-only.yaml:
```diff
@@ -0,0 +1,22 @@
+# RUN: yaml2obj %s | od -v -An -tx1 | FileCheck --ignore-case %s
+
+## Verify that the GOFF end record is correct, with all fields having a value.
```
Can you change this to state that we're verifying that all fields have the value we expect but that the GOFF is still invalid? We have a lot of developers who don't know GOFF and stating that the record "is correct" won't be helpful to them.

kpneal commented on llvm/test/tools/yaml2obj/GOFF/header-only.yaml:
```diff
@@ -0,0 +1,18 @@
+- ModuleHeader:
+    ArchitectureLevel: 1
+    PropertiesLength: 3
+    Properties: 0333ff
```
Are module properties documented anywhere? The only example I've ever seen was GOFF produced by the Binder, and that GOFF had module properties that were english text. Given that, it doesn't seem right to treat this field as an opaque sequence of bytes in the yaml.

jh7370 commented on llvm/test/tools/yaml2obj/GOFF/end-long-name.yaml:
```diff
@@ -0,0 +1,74 @@
+# RUN: awk 'BEGIN {str = sprintf("%47s", "")} {gsub("@FILL@", str); print}' < %s | \
+# RUN:   yaml2obj | od -v -An -tx1 | \
+# RUN:   FileCheck --ignore-case --check-prefix CHECK1 %s
```
`--ignore-case` shouldn't be necessary if you use the right options with `od`?

jh7370 commented on llvm/test/tools/yaml2obj/GOFF/end-long-name.yaml:
```diff
@@ -0,0 +1,74 @@
+# RUN:   FileCheck --ignore-case --check-prefix CHECK4 %s
+
+
+## Verify that the entry name is written correctly over multiple records.
```
Normally we put the comment before the test case run line that it's related to. Indeed, a common pattern in more recent yaml2obj tests is the following:

```
## High-level comment summarising the tests in the file as a whole.
# RUN: General setup code, e.g. creating a test output directory. Not always needed.

## Comment describing individual test case.
# RUN: Test case commands

# CHECK: Check directive(s) for test case

## Second test case comment
# RUN: Second test case commands

# CHECK-2: Check directive(s) for second test case, if needed

...

<Common YAML>
```
Variations on the above where multiple different YAML docs are needed tend to have that YAML at the end of the block of test cases that share it.

MaskRay commented on llvm/test/tools/yaml2obj/GOFF/invalid-record.yaml:
```diff
@@ -0,0 +1,7 @@
+# RUN: not yaml2obj %s 2>&1 | FileCheck %s
+
+# CHECK: YAML:{{[0-9]*}}:{{[0-9]*}}: error: unknown record type name 'Foo'
```
Note, we usually omit the part before `error:`.

Just use `CHECK: error: xxx` (`k` typo below?)
